substrate/frame/revive/src/lib.rs

```diff
@@ -850,34 +991,44 @@ pub mod pallet {
 			#[pallet::compact] storage_deposit_limit: BalanceOf<T>,
 			code: Vec<u8>,
 			data: Vec<u8>,
+			payload: Vec<u8>,
 		) -> DispatchResultWithPostInfo {
-			let code_len = code.len() as u32;
-			let data_len = data.len() as u32;
-			let mut output = Self::bare_instantiate(
-				origin,
-				value,
-				gas_limit,
-				DepositLimit::Balance(storage_deposit_limit),
-				Code::Upload(code),
-				data,
-				None,
-				BumpNonce::No,
-			);
+			ensure_signed(origin.clone())?;
 
-			if let Ok(retval) = &output.result {
-				if retval.result.did_revert() {
-					output.result = Err(<Error<T>>::ContractReverted.into());
+			block_storage::with_ethereum_context(|| {
+				let code_len = code.len() as u32;
+				let data_len = data.len() as u32;
+				let mut output = Self::bare_instantiate(
+					origin,
+					value,
+					gas_limit,
+					DepositLimit::Balance(storage_deposit_limit),
+					Code::Upload(code),
+					data,
+					None,
+					BumpNonce::No,
+				);
+
+				if let Ok(retval) = &output.result {
+					if retval.result.did_revert() {
+						output.result = Err(<Error<T>>::ContractReverted.into());
+					}
 				}
-			}
-			dispatch_result(
-				output.result.map(|result| result.result),
-				output.gas_consumed,
-				T::WeightInfo::eth_instantiate_with_code(
-					code_len,
-					data_len,
-					Pallet::<T>::has_dust(value).into(),
-				),
-			)
+
+				// TODO: Should we report `gas_consumed.saturating_add(base_weight)` instead?
+				// If so, we might need to capture this from inside the `dispatch_result`.
+				Self::store_transaction(payload, output.result.is_ok(), output.gas_consumed);
```

Comment:
this will be added to the weight of the benchmark when we re-run it so this is fine.
later we will also add the little additional weight that needs to be accounted for for looping through all these tx in on_finalize to compute the hashes, but this will done by Lucasz